toffy/fov_watcher.py

```diff
@@ -0,0 +1,217 @@
+import os
+import time
+import json
+from datetime import datetime
+from typing import Callable, List, Union, Tuple
+from watchdog.events import FileCreatedEvent, FileSystemEventHandler
+from watchdog.observers import Observer
+
+import pandas as pd
+import xarray as xr
+
+from mibi_bin_tools import bin_files
+
+class RunStructure:
+    """Expected bin and json files
+
+    Attributes:
+        completion (dict): Whether or not an expected file has been created
+    """
+    def __init__(self, run_folder: str):
+        """ initializes RunStructure by parsing run json within provided run folder
+
+        Args:
+            run_folder (str):
+                path to run folder
+        """
+        self.completion = {}
+        
+        # find run .json and get parameters
+        with open(f'{run_folder}.json', 'r') as f:
+            run_metadata = json.load(f)
+    
+        # parse run_metadata and populate expected structure
+        for fov in run_metadata.get('fovs', ()):
+            run_order = fov.get('runOrder', -1)
+            scan = fov.get('scanCount', -1)
+            if run_order * scan < 0:
+                raise KeyError(f"Could not locate keys in {run_folder}.json")
+            
+            fov_name = f'fov-{run_order}-scan-{scan}'
+            self.completion[fov_name] = {
+                'json': False,
+                'bin': False,
+            }
+
+    def check_run_condition(self, path: str) -> Tuple[bool, str]:
+        """Checks if all requisite files exist and are complete
+
+        Args:
+            path (str):
+                path to expected file
+    
+        Returns:
+            (bool, str):
+                whether or not both json and bin files exist, as well as the name of the point
+        """
+
+        # TODO: check watchdog path depth
+
+        fov_name, extension = path.split('.')[0:2]
+        if fov_name in self.completion:
```

Comment:
If `fov_name` isn't in the `completion` dict we should raise some type of error, since it's initialized with everything from the directory right?

Reply:
right now, it's only checking for bin/json files. I'd say it should only raise an error if `fov_name` isn't a key in `completion`, **and** `extension` is `'bin'`.